x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,1135 @@
+    }
+
+    fn prefetch_extension_information(&self, name: &'static str) -> Fut<'_, (), ConnectionError> {
+        Box::pin(async move {
+            self.prefetech_extension_impl(name).await?;
+            Ok(())
+        })
+    }
+
+    fn extension_information(
+        &self,
+        name: &'static str,
+    ) -> Fut<'_, Option<ExtensionInformation>, ConnectionError> {
+        Box::pin(async move {
+            // Make sure we've prefetched the information.
+            let ext = self.prefetech_extension_impl(name).await?;
+
+            // Complete the request if necessary.
+            if let Some(ExtensionState::Loading(_)) = ext.get(name) {
+                // Acquire a write lock and try again.
+                drop(ext);
+                let mut ext = self.extensions.write().await;
+
+                if let Some(ExtensionState::Loading(seq)) = ext.get(name) {
+                    // We need to wait for the reply.
+                    let cookie = Cookie::<'_, _, QueryExtensionReply>::new(self, *seq);
+                    let reply = cookie.reply().await.map_err(|e| match e {
+                        ReplyError::ConnectionError(e) => e,
+                        ReplyError::X11Error(_) => ConnectionError::UnknownError,
+                    })?;
+
+                    // Mark the extension as loaded.
+                    let _ = ext.insert(name, ExtensionState::Loaded);
+
+                    // Take the "present" field of the reply and use it to determine whether the
+                    // extension is present.
+                    if reply.present {
+                        self.inner.external_insert_extension(
+                            name,
+                            Some(ExtensionInformation {
+                                major_opcode: reply.major_opcode,
+                                first_event: reply.first_event,
+                                first_error: reply.first_error,
+                            }),
+                        )
+                    } else {
+                        self.inner.external_insert_extension(name, None)
+                    }
+                }
+            }
+
+            // Should never block now.
+            self.inner.extension_information(name)
+        })
+    }
+
+    fn maximum_request_bytes(&self) -> Pin<Box<dyn Future<Output = usize> + Send + '_>> {
+        Box::pin(async move {
+            // Make sure we've prefetched the information.
+            let mrl = match self
+                .prefetch_len_impl()
+                .await
+                .expect("prefetch_len_impl failed")
+            {
+                Some(mrl) => mrl,
+                None => {
+                    // We must be writing to it at the moment.
+                    return self
+                        .setup()
+                        .maximum_request_length
+                        .try_into()
+                        .unwrap_or(std::usize::MAX)
+                        * 4;
```

Comment:
This `unwrap_or` seems like a trap waiting to explode. I doubt that `MAX * 4` fits into an usize without overflow. ;-)